### ISSUE TITLE
Support for C++11 strongly typed enumerations breaks the code compatibility

### ISSUE THREAD

**Issue by oliora** [swig/swig#316]
I have a strongly typed enum, something like:

```
enum class BatchType {
    None,
    Generation,
    Export,
};
```

When I used SWIG 3.0.0 to generate Python wrapper for this enum I got the following variables defined:
- `None`
- `Generation`
- `Export`

After switching to SWIG 3.0.4 (which has support for C++11 strongly typed enums) the names have changed to
- `BatchType_None`
- `BatchType_Generation`
- `BatchType_Export`

So my Python code using this enum is broken now.

I think it's better to generate new style code only if it's explicitly requested by command line parameter or whatever, because minor SWIG update should not break the backward compatibility. Or at least there should be a simple way to ask new SWIG to generate backward compatible code for strongly typed enums.

**Comment by ptomulik** [comment]
Hi @oliora, I'm working on #251 which may address this issue. However it needs attention of swig devels, as I've stuck with some issues described therein. IHMO, there should be possibility to keep old enums and introduce new scoping scheme (``BatchType.None`` etc.).

**Comment by oliora** [comment]
@ptomulik agree, it need to be configured. And new enums should be represented in Python as scoped like `BatchType.None` instead of `BatchType_None`.
We have downgrade the SWIG version back to 3.0.0 so this issue is not urgent. But until fixed it could prevent users from upgrading.

**Comment by wsfulton** [comment]
3.0.0 has a bug with the strongly typed enum names as you've pointed out. We aren't going to make it bug compatible to past releases, especially not the first in a major release. You'll need to use the new names. The old names will produce name clashes. This is still work in progress and strongly typed enums require nested class support to be properly implemented first as well as %nspace as they all relate to scoping of Python objects, but until then everything will be flattened in the Python module. I'd like to see a patch for these before strongly typed enums to make sure they fit in nicely with these more important features. Applies to #251 too.

**Comment by ptomulik** [comment]
Nested classes and %nspace are topics that appear again and again over the last few years, but it looks like there is no man so brave to put hands and implement them... Is there any reason for that? :)

**Comment by ojwb** [comment]
William's classified this as a bug in 3.0.0, and looking at CHANGES it was fixed by 3.0.3.  I tend to agree that not trying to be bug compatible was the right call here, but even if it weren't we're approaching 4.1.0 now so for many more versions and much more time we've had the current behaviour compared to what 3.0.0 to 3.0.2 did, therefore closing.
